File: src/main/java/com/anod/appwatcher/backup/gdrive/ApiClientAsyncTask.java
// Copyright 2013 Google Inc. All Rights Reserved.

package com.anod.appwatcher.backup.gdrive;

import android.content.Context;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.GoogleApiClient.ConnectionCallbacks;
import com.google.android.gms.common.api.GoogleApiClient.OnConnectionFailedListener;
import com.google.android.gms.drive.Drive;

import java.util.concurrent.CountDownLatch;

/**
 * An AsyncTask that maintains a connected client.
 */
public abstract class ApiClientAsyncTask<Params, Progress, Result>
        extends AsyncTask<Params, Progress, Result> {

    private GoogleApiClient mClient;
	protected Context mContext;
    public ApiClientAsyncTask(Context context, GoogleApiClient client) {

        mClient = client;
		mContext = context;
    }

    @Override
    protected final Result doInBackground(Params... params) {
        Log.d("TAG", "in background");
        final CountDownLatch latch = new CountDownLatch(1);
        mClient.registerConnectionCallbacks(new ConnectionCallbacks() {
            @Override
            public void onConnectionSuspended(int cause) {
            }

            @Override
            public void onConnected(Bundle arg0) {
                latch.countDown();
            }
        });
        mClient.registerConnectionFailedListener(new OnConnectionFailedListener() {
            @Override
            public void onConnectionFailed(ConnectionResult arg0) {
                latch.countDown();
            }
        });
        mClient.connect();
        try {
            latch.await();
        } catch (InterruptedException e) {
            return null;
        }
        if (!mClient.isConnected()) {
            return null;
        }
        try {
            return doInBackgroundConnected(params);
        } finally {
            mClient.disconnect();
        }
    }

    /**
     * Override this method to perform a computation on a background thread, while the client is
     * connected.
     */
    protected abstract Result doInBackgroundConnected(Params... params);

    /**
     * Gets the GoogleApliClient owned by this async task.
     */
    protected GoogleApiClient getGoogleApiClient() {
        return mClient;
    }
}

File: src/main/java/com/anod/appwatcher/backup/gdrive/ApiClientAsyncTask.kt
// Copyright 2013 Google Inc. All Rights Reserved.

package com.anod.appwatcher.backup.gdrive

import android.os.AsyncTask
import android.os.Bundle
import android.util.Log

import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.GoogleApiClient.ConnectionCallbacks

import java.util.concurrent.CountDownLatch

/**
 * An AsyncTask that maintains a connected client.
 */
abstract class ApiClientAsyncTask(protected val googleApiClient: GoogleApiClient)
    : AsyncTask<Void, Void, ApiClientAsyncTask.Result>() {

    class Result(var status: Boolean, var ex: Exception?)

    override fun doInBackground(vararg params: Void?): Result? {
        Log.d("TAG", "in background")
        val latch = CountDownLatch(1)
        googleApiClient.registerConnectionCallbacks(object : ConnectionCallbacks {
            override fun onConnectionSuspended(cause: Int) {}

            override fun onConnected(arg0: Bundle?) {
                latch.countDown()
            }
        })
        googleApiClient.registerConnectionFailedListener { latch.countDown() }
        googleApiClient.connect()
        try {
            latch.await()
        } catch (e: InterruptedException) {
            return Result(false, e)
        }

        if (!googleApiClient.isConnected) {
            return Result(false, null)
        }
        try {
            return doInBackgroundConnected()
        } finally {
            googleApiClient.disconnect()
        }
    }

    /**
     * Override this method to perform a computation on a background thread, while the client is
     * connected.
     */
    abstract fun doInBackgroundConnected(): ApiClientAsyncTask.Result
}